extensions/Worker.Extensions.CosmosDB/src/CosmosDBConverter.cs

```diff
@@ -0,0 +1,201 @@
+                    {
+                        throw;
+                    }
+                }
+            }
+
+            if (context.Source is CollectionModelBindingData collectionModelBindingData)
+            {
+                if (collectionModelBindingData.ModelBindingDataArray.Any(x => x.Source is Constants.ExtensionName))
+                {
+                    try
+                    {
+                        var collectionResult = await ToTargetTypeCollection(context, context.TargetType, collectionModelBindingData);
+
+                        if (collectionResult is not null && collectionResult is { Count: > 0 })
+                        {
+                            return ConversionResult.Success(collectionResult);
+                        }
+                    }
+                    catch (Exception ex)
+                    {
+                        // TODO: DeserializeObject could throw
+                        Console.WriteLine(ex);
+                    }
+                }
+            }
+
+            return ConversionResult.Unhandled();
+        }
+
+        private async Task<object> ToTargetType(Type targetType, CosmosDBInputAttribute cosmosAttribute) => targetType switch
+        {
+            Type _ when targetType == typeof(CosmosClient) => CreateCosmosClient<CosmosClient>(cosmosAttribute),
+            Type _ when targetType == typeof(Database) => CreateCosmosClient<Database>(cosmosAttribute),
+            Type _ when targetType == typeof(Container) => CreateCosmosClient<Container>(cosmosAttribute),
+            _ => await CreateTargetObject(targetType, cosmosAttribute)
+        };
+
+        private async Task<List<object>> ToTargetTypeCollection(ConverterContext context, Type targetType, CollectionModelBindingData collectionModelBindingData)
+        {
+            var collectionCosmosItems = new List<object>(collectionModelBindingData.ModelBindingDataArray.Length);
+
+            foreach (ModelBindingData modelBindingData in collectionModelBindingData.ModelBindingDataArray)
+            {
+                var cosmosAttribute = modelBindingData.Content.ToObjectFromJson<CosmosDBInputAttribute>();
+                var cosmosItem = await ToTargetType(targetType, cosmosAttribute);
+                if (cosmosItem is not null)
+                {
+                    collectionCosmosItems.Add(cosmosItem);
+                }
+            }
+
+            return collectionCosmosItems;
+        }
+
+        private async Task<object> CreateTargetObject(Type targetType, CosmosDBInputAttribute cosmosAttribute)
+        {
+            // if target type is a collection and NOT of type IList, should we handle this early
+            // and let users know we only support IList types?
+
+            if (targetType.GenericTypeArguments.Any())
+            {
+                targetType = targetType.GenericTypeArguments.FirstOrDefault();
+            }
+
+            MethodInfo createPOCOFromReferenceMethod = GetType()
```

Comment:
Needing to use reflection here is less than ideal. Is it possible to do it without reflection? For example, items can be read without generics via `Container.ReadItemStreamAsync`, which will give the item as a `Stream`, which leaves deserializing as the next challenge. Do we use the serializer from Cosmos? Do we have access to that (I think it is just default JSON)? Or do we use the serializer from `WorkerOptions`? And is that worth it or do we live with the reflection?

I'll leave this up to you. I would like to see the reflection removed as it is always nice to avoid it. But the potential serialization inconsistency may be even worse.

Reply:
Yeah that's a fair point - I think either way handling this whole POCO conversion is going to be difficult :( I can look into another approach but it might be worth a design discussion with the team

Reply:
We are trying to avoid/rewrite reflection code as much as we can to make isolated AOT friendly. #1056 tracks the rewrite needed for existing code.